### DISCUSSION TITLE
Why eager mode need memcpy on receiver side?

### DISCUSSION THREAD

**Discussion by haipeng31** [openucx/ucx#7243] [Q&A] [answered]
In my opinion, ibv_post_recv can post a registered memory,  when HCA receive data, it can access the buffer directly.

**Comment by yosefe** [comment, marked as answer]
eager mode is modeled after MPI tag-matching semantics, where the data has to arrive to a specific buffer according to the tag.
on other hand, with UCP active message API, it's possible to access the data directly from the callback without copy:
https://openucx.readthedocs.io/en/master/api.html#_CPPv417ucp_am_callback_t

**Reply by haipeng31** [reply]
thanks for you reply, it helps a lot.
